[CSSolver] Fix a crash in diagnostics related to pattern matching

xedin opened the pull request:
`repairFailures` code for sequence element type attempts to assign placeholders to pattern type directly but it didn't check whether type variables that appear in it are already bound (`repairFailures` is called from `matchTypes` that doesn't fully simplify involved types). Adjusting that fixes the crash but results in sub-par diagnostic because type of `is` pattern wasn't allowed to be bound to a hole which means that placeholders cannot be propagated from the sequence element type down into associated pattern in this case.

Resolves: https://github.com/apple/swift/issues/65360

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[CSSimplify] Fix attempt to rebind type variables in sequence element pattern

xedin committed:
[CSGen] Allow `is` pattern types to be holes

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

hamishknight commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -6161,8 +6161,10 @@ bool ConstraintSystem::repairFailures(
     if (lhs->isAny()) {
       rhs.visit([&](Type type) {
         if (auto *typeVar = type->getAs<TypeVariableType>()) {
-          assignFixedType(typeVar,
-                          PlaceholderType::get(getASTContext(), typeVar));
+          if (!getFixedType(typeVar)) {
+            assignFixedType(typeVar,
+                            PlaceholderType::get(getASTContext(), typeVar));
+          }
```
Would it be better to `simplifyType(rhs)` here? Mainly asking as this pattern is used in a couple of other places:

https://github.com/apple/swift/blob/1c3667b52aa7888f7e2cc253b725a7ed055e4ef2/lib/Sema/CSStep.cpp#L1106-L1130

https://github.com/apple/swift/blob/1c3667b52aa7888f7e2cc253b725a7ed055e4ef2/lib/Sema/BuilderTransform.cpp#L1178-L1182

And I've just realized there's a very subtle footgun to their implementation: because `getAs` looks through sugar, if it visits e.g a ParenType, then it would end up trying to `assignFixedType` twice, which would blow up. Maybe we should factor out this logic into a function that ensures this doesn't happen?

xedin replied:
`simplifyType` transforms which I am generally cautious about, that’s why I didn’t use it here, and in general we shouldn’t be using `assignFixedType` directly too. I considered to use Bind constraint instead but my worry is that the flags are not always reliable to indicate whether new constraints are allowed…

hamishknight replied:
Yeah, I'm just wondering if there's something we can factor out to avoid the footgun with `getAs`. Maybe we should at least change them to use `dyn_cast` for now to avoid potential issues (I haven't yet tried to come up with cases that would hit the issue, but it seems like it should be possible).

xedin replied:
`getFixedType` is the only way to go I guess. I'm going to extract this into a function.

xedin replied:
Alright, done!

hamishknight replied:
Thanks!

xedin committed:
[ConstraintSystem] NFC: Standardize the way of type holefication

Introduce `ConstraintSystem::recordTypeVariablesAsHoles` as a
standard way to record that unbound type variables found in a
type are holes in the given constraint system.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

AnthonyLatsis commented on validation-test/Sema/type_checker_crashers_fixed/issue-65360.swift:
```diff
@@ -0,0 +1,8 @@
+let _: () -> Void = {
+  for case (is Int)? in [a] {}
+  if case (is Int, is Int) = a {} // expected-error {{cannot convert value of type 'Any?' to specified type '(_, _)'}}
+}
```
Ideas for a few more test cases:

```swift
let _: () -> Void = {
  for case (0)? in [a] {}
  if case (0, 0) = a {}
}

let _: () -> Void = {
  for case (0)? in [a] {}
  for case (0, 0) in [a] {}
}
```

xedin replied:
Added, thank you!

AnthonyLatsis reviewed (commented):
Whoa, super quick fix!

xedin committed:
[CSGen] Allow expression pattern types to be holes

xedin commented:
@swift-ci please test
